File: app/src/main/java/com/song/trust/applist/ApplicationBean.java
package com.song.trust.applist;

import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;

import org.jetbrains.annotations.NotNull;

/**
 * Created by chensongsong on 2021/11/19.
 */
public class ApplicationBean implements Parcelable, Comparable<ApplicationBean> {

    private String name;
    private String packageName;
    private Drawable icon;
    private boolean configured;
    private String title;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPackageName() {
        return packageName;
    }

    public void setPackageName(String packageName) {
        this.packageName = packageName;
    }

    public Drawable getIcon() {
        return icon;
    }

    public void setIcon(Drawable icon) {
        this.icon = icon;
    }

    public boolean isConfigured() {
        return configured;
    }

    public void setConfigured(boolean configured) {
        this.configured = configured;
    }

    public String getTitle() {
        return title;
    }

    public ApplicationBean setTitle(String title) {
        this.title = title;
        return this;
    }

    @Override
    public int compareTo(ApplicationBean applicationBean) {
        if (this.isConfigured() == applicationBean.isConfigured()) {
            return 0;
        } else if (this.isConfigured()) {
            return -1;
        } else {
            return 1;
        }
    }

    @NotNull
    @Override
    public String toString() {
        return "ApplicationBean{" +
                "name='" + name + '\'' +
                ", packageName='" + packageName + '\'' +
                ", configured=" + configured +
                ", title='" + title + '\'' +
                '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.name);
        dest.writeString(this.packageName);
        dest.writeByte(this.configured ? (byte) 1 : (byte) 0);
        dest.writeString(this.title);
    }

    public ApplicationBean() {
    }

    protected ApplicationBean(Parcel in) {
        this.name = in.readString();
        this.packageName = in.readString();
        this.configured = in.readByte() != 0;
        this.title = in.readString();
    }

    public static final Creator<ApplicationBean> CREATOR = new Creator<ApplicationBean>() {
        @Override
        public ApplicationBean createFromParcel(Parcel source) {
            return new ApplicationBean(source);
        }

        @Override
        public ApplicationBean[] newArray(int size) {
            return new ApplicationBean[size];
        }
    };
}

File: app/src/main/java/com/song/trust/applist/ApplicationBean.kt
package com.song.trust.applist

import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable

/**
 * Created by chensongsong on 2021/11/19.
 */
open class ApplicationBean : Parcelable, Comparable<ApplicationBean> {
    var name: String? = null
    var packageName: String? = null
    var icon: Drawable? = null
    var isConfigured = false
    var title: String? = null
        private set

    fun setTitle(title: String?): ApplicationBean {
        this.title = title
        return this
    }

    override fun compareTo(other: ApplicationBean): Int {
        return when {
            isConfigured == other.isConfigured -> {
                0
            }
            isConfigured -> {
                -1
            }
            else -> {
                1
            }
        }
    }

    override fun toString(): String {
        return "ApplicationBean{" +
                "name='" + name + '\'' +
                ", packageName='" + packageName + '\'' +
                ", configured=" + isConfigured +
                ", title='" + title + '\'' +
                '}'
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeString(packageName)
        dest.writeByte(if (isConfigured) 1.toByte() else 0.toByte())
        dest.writeString(title)
    }

    constructor()
    protected constructor(`in`: Parcel) {
        name = `in`.readString()
        packageName = `in`.readString()
        isConfigured = `in`.readByte().toInt() != 0
        title = `in`.readString()
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<ApplicationBean?> =
            object : Parcelable.Creator<ApplicationBean?> {
                override fun createFromParcel(source: Parcel): ApplicationBean {
                    return ApplicationBean(source)
                }

                override fun newArray(size: Int): Array<ApplicationBean?> {
                    return arrayOfNulls(size)
                }
            }
    }
}